README.md
| Latest Version | PowerShell Gallery | Issues | License | Discord |
|-----------------|----------------|----------------|----------------|----------------|
| [![Latest Version](https://img.shields.io/github/v/tag/PoshAdoTasks/PoshAdoTask)](https://github.com/PoshAdoTasks/PoshAdoTask/tags) | [![Powershell Gallery](https://img.shields.io/powershellgallery/dt/PoshAdoTask)](https://www.powershellgallery.com/packages/PoshAdoTask) | [![GitHub issues](https://img.shields.io/github/issues/PoshAdoTasks/PoshAdoTask)](https://github.com/PoshAdoTasks/PoshAdoTask/issues) | [![GitHub license](https://img.shields.io/github/license/PoshAdoTasks/PoshAdoTask)](https://github.com/PoshAdoTasks/PoshAdoTask/blob/master/LICENSE) | [![Discord Server](https://assets-global.website-files.com/6257adef93867e50d84d30e2/636e0b5493894cf60b300587_full_logo_white_RGB.svg)](https://discord.com/channels/1044305359021555793/1044305460729225227) |
# PoshAdoTask Module

## Description

This module provides a set of tools for build Azure Devops Extensions.

## PoshAdoTask Cmdlets

### [Add-Contribution](Docs/Add-Contribution.md)

Add a contribution to the Manifest

### [Add-File](Docs/Add-File.md)

Add a file to the Manifest

### [Add-Input](Docs/Add-Input.md)

Add an input to the Task

### [Add-Target](Docs/Add-Target.md)

Add a target to the Manifest

### [Get-Manifest](Docs/Get-Manifest.md)

Get the Manifest

### [Get-Task](Docs/Get-Task.md)

Get the Task

### [New-Content](Docs/New-Content.md)

Add content to the Manifest

### [New-Execution](Docs/New-Execution.md)

Add an execution to the Task

### [New-Icon](Docs/New-Icon.md)

Add an icon to the Manifest

### [New-Link](Docs/New-Link.md)

Add a link to the Manifest

### [New-Manifest](Docs/New-Manifest.md)

Create the Manifest

### [New-Repository](Docs/New-Repository.md)

Add a repository to the Manifest

### [New-Task](Docs/New-Task.md)

Create the Task

### [Out-Manifest](Docs/Out-Manifest.md)

Write the Manifest to the console

### [Out-Task](Docs/Out-Task.md)

Write the Task to the console

### [Remove-Contribution](Docs/Remove-Contribution.md)

Remove a contribution from the Manifest

### [Remove-File](Docs/Remove-File.md)

Remove a file from the Manifest

### [Remove-Target](Docs/Remove-Target.md)

Remove a target from the Manifest

### [Set-Category](Docs/Set-Category.md)

Add a Category to the Manifest or Task

### [Set-GalleryFlag](Docs/Set-GalleryFlag.md)

Add a galleryflag to the Manifest

### [Set-Visibility](Docs/Set-Visibility.md)

Add visibility to the Task



CHANGELOG.md
# Changelog

All changes to this module should be reflected in this document.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [[1.1.0]](https://github.com/PoshAdoTasks/PoshAdoTask/releases/tag/v1.1.0) - 2023-01-10

This release fixes some issues discovered while attempting to use the module in production.

Issues Addressed:

1. tfx cli is case-sensitive, manifest.repository.type git <> Git (issue #3)
   1. Changed Repository.Type to String
   2. Updated New-Repository to pass in lowercase cvs,git,mercurial,svn
2. Contribution.Property.Name is not arbitrary (issue #4)
   1. Updated the Add-Contribution to accept -Name
3. Manifest.Version is a string, Task.Version is essentially a System.Version object
   1. Created a new cmdlet to set the version (New-Version)

--

## [[1.0.0]](https://github.com/PoshAdoTasks/PoshAdoTask/releases/tag/v1.0.0) - 2023-01-09

This is the initial release of this module, the goal is to provide a set of tools for working with and developing Azure Devops Extensions as quickly and easily as possible. There are two Types that we work with Manifest and Task, each of these are used in developing the extension.

You can find the documentation for the Extension Manifest [here](https://learn.microsoft.com/en-us/azure/devops/extend/develop/manifest?view=azure-devops#discoveryprops)

You can find the documentation for the Task Refernce [here](https://github.com/Microsoft/azure-pipelines-task-lib/blob/master/tasks.schema.json)

From there I built out an initial set of cmdlets based on what you would find in a typical vss-extension.json and task.json.

--


psakefile.ps1
$script:moduleName = "PoshAdoTask"
$script:Github = "https://github.com/PoshAdoTasks/PoshAdoTask"
$script:PoshGallery = "https://www.powershellgallery.com/packages/PoshAdoTask"
$script:Nuget = ""
$script:DiscordChannel = "https://discord.com/channels/1044305359021555793/1044305460729225227"

Task default -depends UpdateReadme

Task LocalUse -Description "Setup for local use and testing" -depends CreateModuleDirectory, CleanModuleDirectory, CleanProject, BuildProject, CopyModuleFiles -Action {
}

Task UpdateHelp -Description "Update the help files" -depends CreateModuleDirectory, CleanProject, BuildProject, CopyModuleFiles -Action {

 Import-Module -Name ".\Module\$($script:moduleName)" -Scope Global -force;
 New-MarkdownHelp -Module PoshAdoTask -AlphabeticParamsOrder -UseFullTypeName -WithModulePage -OutputFolder .\Docs\ -ErrorAction SilentlyContinue
 Update-MarkdownHelp -Path .\Docs\ -AlphabeticParamsOrder -UseFullTypeName
}

Task UpdateReadme -Description "Update the README file" -depends CreateModuleDirectory, CleanProject, BuildProject, CopyModuleFiles -Action {

 $readMe = Get-Item .\README.md

 $TableHeaders = "| Latest Version | PowerShell Gallery | Issues | License | Discord |"
 $Columns = "|-----------------|----------------|----------------|----------------|----------------|"
 $VersionBadge = "[![Latest Version](https://img.shields.io/github/v/tag/PoshAdoTasks/PoshAdoTask)](https://github.com/PoshAdoTasks/PoshAdoTask/tags)"
 $GalleryBadge = "[![Powershell Gallery](https://img.shields.io/powershellgallery/dt/PoshAdoTask)](https://www.powershellgallery.com/packages/PoshAdoTask)"
 $IssueBadge = "[![GitHub issues](https://img.shields.io/github/issues/PoshAdoTasks/PoshAdoTask)](https://github.com/PoshAdoTasks/PoshAdoTask/issues)"
 $LicenseBadge = "[![GitHub license](https://img.shields.io/github/license/PoshAdoTasks/PoshAdoTask)](https://github.com/PoshAdoTasks/PoshAdoTask/blob/master/LICENSE)"
 $DiscordBadge = "[![Discord Server](https://assets-global.website-files.com/6257adef93867e50d84d30e2/636e0b5493894cf60b300587_full_logo_white_RGB.svg)]($($script:DiscordChannel))"

 if (!(Get-Module -Name $script:moduleName )) { Import-Module -Name ".\Module\$($script:moduleName).psd1" }

 Write-Output $TableHeaders | Out-File $readMe.FullName -Force
 Write-Output $Columns | Out-File $readMe.FullName -Append
 Write-Output "| $($VersionBadge) | $($GalleryBadge) | $($IssueBadge) | $($LicenseBadge) | $($DiscordBadge) |" | Out-File $readMe.FullName -Append

 Get-Content .\Docs\PoshAdoTask.md | Select-Object -Skip 8 | ForEach-Object { $_.Replace('(', '(Docs/') } | Out-File $readMe.FullName -Append
 Write-Output "" | Out-File $readMe.FullName -Append
 #Get-Content .\Build.md | Out-File $readMe.FullName -Append
}

Task SetupModule -Description "Setup the PowerShell Module" -depends CreateModuleDirectory, CleanModuleDirectory, CleanProject, BuildProject, CopyModuleFiles, CreateExternalHelp, CreateCabFile, CreateNuSpec, NugetPack, NugetPush

Task NewTaggedRelease -Description "Create a tagged release" -depends CreateModuleDirectory, CleanProject, BuildProject, CopyModuleFiles -Action {


 if (!(Get-Module -Name $script:moduleName )) { Import-Module -Name ".\Module\$($script:moduleName).psd1" }
 $Version = (Get-Module -Name $script:moduleName | Select-Object -Property Version).Version.ToString()
 git tag -a v$version -m "$($script:moduleName) Version $($Version)"
 git push origin v$version
}

Task Post2Discord -Description "Post a message to discord" -Action {

 $version = (Get-Module -Name $script:moduleName | Select-Object -Property Version).Version.ToString()
 $Discord = Get-Content .\discord.PoshAdoTask | ConvertFrom-Json
 $Discord.message.content = "Version $($version) of $($script:moduleName) released. Please visit Github ($($script:Github)) or PowershellGallery ($($script:PoshGallery)) to download."
 Invoke-RestMethod -Uri $Discord.uri -Body ($Discord.message | ConvertTo-Json -Compress) -Method Post -ContentType 'application/json; charset=UTF-8'
}

Task CleanProject -Description "Clean the project before building" -Action {
 dotnet clean .\PoshAdoTask\PoshAdoTask.csproj
}

Task BuildProject -Description "Build the project" -Action {
 dotnet build .\PoshAdoTask\PoshAdoTask.csproj -c Release
}

Task CreateModuleDirectory -Description "Create the module directory" -Action {
 New-Item Module -ItemType Directory -Force
}

Task CleanModuleDirectory -Description "Clean the module directory" -Action {
 Get-ChildItem .\Module\ | Remove-Item
}

Task CopyModuleFiles -Description "Copy files for the module" -Action {
 Copy-Item .\PoshAdoTask\bin\Release\net6.0\*.dll Module -Force
 Copy-Item .\PoshAdoTask.psd1 Module -Force
}

Task CreateExternalHelp -Description "Create external help file" -Action {
 New-ExternalHelp -Path .\Docs -OutputPath .\Module\ -Force
}

Task CreateCabFile -Description "Create cab file for download" -Action {
 New-ExternalHelpCab -CabFilesFolder .\Module\ -LandingPagePath .\Docs\PoshAdoTask.md -OutputFolder .\cabs\
}

Task CreateNuSpec -Description "Create NuSpec file for upload" -Action {
 .\ConvertTo-NuSpec.ps1 -ManifestPath ".\Module\PoshAdoTask.psd1"
}

Task NugetPack -Description "Pack the nuget file" -Action {
 nuget pack .\Module\PoshAdoTask.nuspec -OutputDirectory .\Module
}

Task NugetPush -Description "Push nuget to PowerShell Gallery" -Action {
 $config = [xml](Get-Content .\nuget.config)
 nuget push .\Module\*.nupkg -NonInteractive -ApiKey "$($config.configuration.apikeys.add.value)" -ConfigFile .\nuget.config
}

Task PesterTest -Description "Test Cmdlets" -Action {
 Invoke-Pester
}